Allow `use super::*;` glob imports

MrAwesome opened the pull request:
changelog: Allow super::* glob imports

fixes #5554 
fixes #5569

A first pass at #5554 - this allows all `use super::*` to pass, which may or may not be desirable. The original issue was around allowing test modules to import their entire parent modules - I'm happy to modify this to do that instead, may just need some guidance on how to implement that (I played around a bit with #[cfg(test)] but from what I can gather, clippy itself isn't in test mode when running, even if the code in question is being checked for the test target).

flip1995 commented:
You cannot check for the `cfg(test)` attribute, since `cfg`/`cfg_attr`s don't exist anymore in the HIR.

dylni commented:
Since `wildcard_imports` is starting to warn in less cases, should it have a configuration option to always warn? That would also help make the [documentation](https://rust-lang.github.io/rust-clippy/master/index.html#wildcard_imports) clearer about it not warning in all cases by default.

flip1995 commented:
We could add this, yes

MrAwesome commented:
Updated to check for "test" in the parent module name. I ended up doing this inline for simplicity's sake, happy to set a field on WildcardImports if that seems better.

flip1995 commented:
> I ended up doing this inline for simplicity's sake, happy to set a field on WildcardImports if that seems better.

That was my first idea when I reviewed this PR too. But that doesn't work, if the super import isn't directly in the main module. And allowing `use super::*` in test functions seems reasonable too. [Playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=0a611006689f73e0c3f70fa056b105d9) for test case.

flip1995 commented:
Better than a bool would be to add a counter how many test modules deep the lint check is, so it doesn't start linting in test modules again, just because a submodule is also called `*test*`.

MrAwesome commented:
Updated to check for "test" in parent module/function names. 

@dylni I also went ahead and added the flag for #5569 since I'm already warmed up on the code.

flip1995 commented on clippy_lints/src/wildcard_imports.rs:
```diff
@@ -73,18 +73,38 @@ declare_clippy_lint! {
+#[derive(Default)]
+pub struct WildcardImports {
+    warn_on_all: bool,
+    is_test_module: bool,
```
This should not be necessary with the field below.

flip1995 commented on clippy_lints/src/wildcard_imports.rs:
```diff
@@ -152,6 +172,19 @@ impl LateLintPass<'_, '_> for WildcardImports {
     }
+
+    fn check_item_post(&mut self, _: &LateContext<'_, '_>, _: &Item<'_>) {
+        if self.is_test_module {
```
```suggestion
        if is_test_module(item) {
```

flip1995 commented on clippy_lints/src/wildcard_imports.rs:
```diff
@@ -168,9 +201,6 @@ fn is_super_only_import(segments: &[PathSegment<'_>]) -> bool {
-    let parent_name = parent_item.ident.name.as_str();
-    parent_name.contains("test")
+fn is_test_module(item: &Item<'_>) -> bool {
+    item.ident.name.as_str().contains("test")
```
You also have to check the `ItemKind` here.

MrAwesome commented:
Okay, updated to remove that field and check the itemkind. I checked for either `fn` or `mod`, should I do just `mod`? 

Also - are there docs I should update as well?

dylni commented:
@MrAwesome Thanks! It's great to see it will be available immediately with this PR.

flip1995 commented:
Yes please exclude `Fn` here, since `test` is a pretty common string in a method name, like `test_for_xyz`, IMO.

MrAwesome commented:
Updated to use Mod only, reflected that in the tests, and updated the docs to reflect all the new functionality.

flip1995 commented on clippy_lints/src/wildcard_imports.rs:
```diff
@@ -73,18 +78,35 @@ declare_clippy_lint! {
-            if !in_macro(item.span);
+            if self.warn_on_all || !in_macro(item.span);
             if let ItemKind::Use(use_path, UseKind::Glob) = &item.kind;
-            // don't lint prelude glob imports
-            if !use_path.segments.iter().last().map_or(false, |ps| ps.ident.as_str() == "prelude");
+            if self.warn_on_all || !self.check_exceptions(use_path.segments);
```
duplicated

MrAwesome replied:
Ah, ok - wasn't sure if the in_macro check still needed to live above the itemkind check.

flip1995 commented on clippy_lints/src/wildcard_imports.rs:
```diff
@@ -73,18 +78,35 @@ declare_clippy_lint! {
         if item.vis.node.is_pub() || item.vis.node.is_pub_restricted() {
             return;
         }
+        if is_test_module_or_function(item) {
+            self.test_modules_deep += 1;
+        }
```
You have to move this to the top of the function.

flip1995 commented on clippy_lints/src/wildcard_imports.rs:
```diff
@@ -153,4 +174,34 @@ impl LateLintPass<'_, '_> for WildcardImports {
+
+    fn check_item_post(&mut self, _: &LateContext<'_, '_>, item: &Item<'_>) {
+        if is_test_module_or_function(item) {
+            self.test_modules_deep -= 1;
```
To be save you can use `saturating_{sub,add}` instead.

MrAwesome commented:
This look correct? Most of my experience is with Mercurial so I want to make extra sure I don't cause some weird repo state.

flip1995 commented:
Perfect, thanks!

@bors r+

bors commented:
:pushpin: Commit b69200b8468434bc3f5b9ef8468733e5d40f4e01 has been approved by `flip1995`

<!-- @bors r=flip1995 b69200b8468434bc3f5b9ef8468733e5d40f4e01 -->
<!-- homu: {"type":"Approved","sha":"b69200b8468434bc3f5b9ef8468733e5d40f4e01","approver":"flip1995"} -->

bors commented:
:hourglass: Testing commit b69200b8468434bc3f5b9ef8468733e5d40f4e01 with merge 43a1777b89cf6791f9e20878b4e5e3ae907867a5...
<!-- homu: {"type":"BuildStarted","head_sha":"b69200b8468434bc3f5b9ef8468733e5d40f4e01","merge_sha":"43a1777b89cf6791f9e20878b4e5e3ae907867a5"} -->

bors commented:
:sunny: Test successful - [checks-action_dev_test](https://github.com/rust-lang/rust-clippy/runs/659452572), [checks-action_remark_test](https://github.com/rust-lang/rust-clippy/runs/659452153), [checks-action_test](https://github.com/rust-lang/rust-clippy/runs/659465669)
Approved by: flip1995
Pushing 43a1777b89cf6791f9e20878b4e5e3ae907867a5 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"flip1995","base_ref":"master","builders":{"checks-action_test":"https://github.com/rust-lang/rust-clippy/runs/659465669","checks-action_dev_test":"https://github.com/rust-lang/rust-clippy/runs/659452572","checks-action_remark_test":"https://github.com/rust-lang/rust-clippy/runs/659452153"},"merge_sha":"43a1777b89cf6791f9e20878b4e5e3ae907867a5"} -->
